src/psyclone/tests/psyir/transformations/loop_tiling_2d_trans_test.py

```diff
@@ -0,0 +1,166 @@
+            integer:: i, j, a
+            integer, intent(inout), dimension(100,100) :: tmp
+
+            do i=1, 100
+              a = 4
+              do j=1, 100
+                tmp(i,j) = 2 * tmp(i,j)
+              enddo
+            enddo
+        end subroutine test
+     ''')
+    outer_loop = psyir.walk(Loop)[0]
+    with pytest.raises(TransformationError) as err:
+        LoopTiling2DTrans().validate(outer_loop)
+    assert ("must be a sub-class of Loop but got 'Assignment'."
+            in str(err.value))
+
+
+def test_loop_tiling_2d_trans_validation2(fortran_reader):
+    ''' Validation fails if the loop body is not another loop. '''
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine test(tmp)
+            integer:: i, j
+            integer, intent(inout), dimension(100,100) :: tmp
+
+            do i=1, 100
+                tmp(i,j) = 2 * tmp(i,j)
+            enddo
+        end subroutine test
+     ''')
+    outer_loop = psyir.walk(Loop)[0]
+    with pytest.raises(TransformationError) as err:
+        LoopTiling2DTrans().validate(outer_loop)
+    assert ("must be a sub-class of Loop but got 'Assignment'."
+            in str(err.value))
+
+
+def test_loop_tiling_2d_trans_validation3(fortran_reader):
+    ''' Validation fails if the loop body may have side effects. '''
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine test(tmp)
+            use other_mod
+            integer:: i, j
+            integer, intent(inout), dimension(100,100) :: tmp
+
+            do i=1, 100
+              do j=1, 100
+                call my_sub(i, j)
+              enddo
+            enddo
+        end subroutine test
+     ''')
+    outer_loop = psyir.walk(Loop)[0]
+    with pytest.raises(TransformationError) as err:
+        LoopTiling2DTrans().validate(outer_loop)
+    assert ("Nodes of type 'Call' cannot be enclosed by a LoopSwapTrans "
+            "transformation" in str(err.value))
+
+
+def test_loop_tiling_2d_trans_apply(fortran_reader, fortran_writer):
+    ''' Validation passes when found a 2D nested loop construct. '''
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine test(tmp)
+            integer:: i, j
+            integer, intent(inout), dimension(100,100) :: tmp
+
+            do i=1, 100
+              do j=1, 100
+                tmp(i,j) = 2 * tmp(i,j)
+              enddo
+            enddo
+        end subroutine test
+     ''')
+    outer_loop = psyir.walk(Loop)[0]
+    LoopTiling2DTrans().apply(outer_loop)
+
+    outer_loop = psyir.walk(Loop)[0]
+    result = fortran_writer(outer_loop)
+    print(result)
+    expected = '''\
+do i_out_var = 1, 100, 32
+  i_el_inner = MIN(i_out_var + (32 - 1), 100)
+  do j_out_var = 1, 100, 32
+    do i = i_out_var, i_el_inner, 1
+      j_el_inner = MIN(j_out_var + (32 - 1), 100)
+      do j = j_out_var, j_el_inner, 1
+        tmp(i,j) = 2 * tmp(i,j)
+      enddo
+    enddo
+  enddo
+enddo'''
+    assert expected in result
```

Comment:
Please add a test that the tilesize can be specified (I believe this won't work currently).

Reply:
Added new test